Userland/Libraries/LibPDF/Filter.cpp

```diff
@@ -225,9 +230,32 @@ PDFErrorOr<ByteBuffer> Filter::handle_lzw_and_flate_parameters(ByteBuffer buffer
     return decode_png_prediction(buffer, bytes_per_row);
 }
 
-PDFErrorOr<ByteBuffer> Filter::decode_lzw(ReadonlyBytes)
+PDFErrorOr<ByteBuffer> Filter::decode_lzw(ReadonlyBytes bytes, int predictor, int columns, int colors, int bits_per_component, int early_change)
 {
-    return Error::rendering_unsupported_error("LZW Filter is not supported");
+    auto memory_stream = make<FixedMemoryStream>(bytes);
+    auto lzw_stream = make<BigEndianInputBitStream>(MaybeOwned<Stream>(move(memory_stream)));
+    Compress::LZWDecoder lzw_decoder { MaybeOwned<BigEndianInputBitStream> { move(lzw_stream) }, 8, -early_change };
+
+    ByteBuffer decoded;
+
+    u16 const clear_code = lzw_decoder.add_control_code();
+    u16 const end_of_data_code = lzw_decoder.add_control_code();
+
+    while (true) {
+        auto const code = TRY(lzw_decoder.next_code());
+
+        if (code == clear_code) {
+            lzw_decoder.reset();
+            continue;
+        }
+
+        if (code == end_of_data_code)
+            break;
+
+        TRY(decoded.try_append(lzw_decoder.get_output()));
```

Comment:
This loop is the same as in LibTiff, right? Should there be a `LZWDecoder::decompress_all()` that has this loop? (Similar to DeflateDecompressor.)

Reply:
If I understand correctly, control codes are not specified by the original LZW paper, but all user (that I know) use the same ones.
And also, GIF expect results output code per output code, TIFF byte per byte and PDF a whole decoded stream.
So it's definitely factorizable, but let's do that later.